WIP Pluggable crypto galore.

cpu opened the pull request:
Continued development of [djc/crypto](https://github.com/djc/rustls/tree/crypto).

djc committed:
crypto: parametrize config types with crypto provider

djc committed:
crypto: redefine single DER constant

djc committed:
crypto: use crypto provider for constant-time comparisons

djc committed:
crypto: use crypto provider for random bytes

djc committed:
crypto: use crypto provider to produce ticketer

djc committed:
kx: let KeyExchange::choose() instantiate directly

cpu committed:
kx: replace pubkey field w/ pub_key() accessor.

The only consumers of the `pub(crate)` visible `pubkey` field of the
`KeyExchange` struct were using it to get at a `&[u8]` of public key
bytes.

This commit:

1. Unexports the `pubkey` field of the `KeyExchange` struct.
2. Adds a `pub(crate)` visible `pub_key()` method to return the public
   key as a `&[u8]`.
3. Adjusts the tls12 client `emit_clientkx` function to use `&[u8]` for
   its pub key argument.
4. Adjusts all callers to use the new `pub_key` accessor in place of the
   field.

The name is changed from `pubkey` to `pub_key` to match Rust naming
conventions[0].

[0]: https://rust-lang.github.io/api-guidelines/naming.html

cpu committed:
crypto: separate module dir, ring sub-module.

For better code organization this commit moves the generic crypto
interface code from `src/crypto.rs` to `src/crypto/lib.rs`.

The *ring* specific code implementing the generic interfaces is moved to
`src/crypto/ring.rs` as a sub-module of `crypto. All imports are
adjusted accordingly.

This has the advantage of leaving `src/crypto/lib.rs` small, and without
any *ring* specific imports. In the future we may choose to feature-gate
the ring sub-module to allow building the crate without a dependency on
ring.

cpu committed:
crypto: add SupportedGroup trait.

This commit adds a trait for referring to supported key exchanges over
named groups in a general fashion. The *ring* specific
`SupportedKxGroup` type is then made to implement this trait.

cpu committed:
crypto: introduce key exchange traits.

This commit adds a `KeyExchange` associated type to the `CryptoProvider`
trait. The `KeyExchange` type is constrained with its own `KeyExchange`
trait that has an associated type for the `SupportedGroup`.

In the `crypto::ring` package we adapt the existing *ring* specific
`KeyExchange` and `SupportedKxGroup` types to these new traits.

Throughout the codebase we tighten generic bounds where required to
ensure we have a `CryptoProvider` bound that allows accessing the
associated `KeyExchange` and `SupportedGroup`. We also make the
`CryptoProvider` an associated type on the `Side` config.

cpu commented:
@djc I think this branch should be a decent improvement over https://github.com/cpu/rustls/pull/2 As requested the PR includes all of the WIP crypto backend commits, not just the key exchange specific bits.

Highlights:
* Rebased again.
* No more distracting import shuffles.
* Smaller commits (as best I could manage).
* Restored the `pub mod kx_group` in `lib.rs`
* Removed the generic parameter on `tls12/mod.rs` `ConnectionSecrets`.
* Less verbose imports in example/helper code.
* Removed a unnecessary `as_slice`.
* Put a `Sized` bound on the `KeyExchange` trait instead of the `start` associated func.
* Renamed `pubkey` to `pub_key`.
* Dropped `Clone` bound on `SupportedGroup` trait.
* Removed stutter in Ring specific type names.
* Renamed `KeyExchange` `all_supported_groups` to `all_kx_groups`.
* Rearranged trait methods to match preferred order.
* Removed explicit var type for turbo-fish use.

The main thing I'm aware of that is still crummy is the `Debug` trait handling required in `builder.rs`. This comes up primarily in 7ab601b23600a804bc1b541c5e3ccf5409a80c3b (helper work) and 418a518d313ae2863337daba1a7acf3f27f7a5fa (gross debug impls). 

I needed to make two of the state `structs` carry forward a reference to the `CryptoProvider` generic parameter so that later states can refer to it, but neither `struct` directly depend on that generic parameter, so a `PhantomData` field is required. That in turn breaks the expected debug output test (`test_config_builders_debug` in `rustls/tests/api.rs`) because the `PhantomData` is included in the debug output. It also requires deriving `Debug` for `crypto::ring::Ring` and `crypto::ring::KeyExchange`. I did my best to find a solution that passes testing, but _there must be a better way_. Maybe there's a way to get the `CryptoProvider` parameter from the `Side` in the deeper states that I need to refer to it from?

vvv committed:
CONTRIBUTING.md: fix a broken link

Jarema committed:
Bump webpki to 0.100.0

Signed-off-by: Tomasz Pietrek <tomasz@nats.io>

djc commented on rustls/src/client/tls13.rs:
```diff
@@ -188,19 +188,22 @@ pub(super) fn initial_key_share(
     server_name: &ServerName,
 ) -> Result<kx::KeyExchange, Error> {
-    let group = match config
+    let group = config
```
Probably good to squash this into the originating commit?

djc commented on rustls/src/kx.rs:
```diff
@@ -55,6 +55,11 @@ impl KeyExchange {
     }
 
+    /// Return the public key being used.
+    pub(crate) fn pub_key(&self) -> &[u8] {
```
Style nit: I would usually put this at the bottom of the impl block (trivial accessor, `&self` receiver).

djc commented on rustls/src/kx.rs:
```diff
@@ -11,8 +11,8 @@ use ring::rand::SystemRandom;
 /// our private key, and our public key.
 pub(crate) struct KeyExchange {
     skxg: &'static SupportedKxGroup,
```
Maybe also rename `skxg`, which is pretty inscrutable? I think it could just be called `group` in this context.

djc commented on examples/src/bin/limitedclient.rs:
```diff
@@ -6,7 +6,7 @@ use std::sync::Arc;
 use std::net::TcpStream;
 
-use rustls::crypto::Ring;
+use rustls::crypto::ring::Ring;
```
This is a good idea, but (contrary to what we did in Quinn) I would move `crypto.rs` to `crypto/mod.rs` to keep it closer to the other files.

djc commented on rustls/src/crypto/ring.rs:
```diff
@@ -0,0 +1,96 @@
+use crate::crypto::CryptoProvider;
+use crate::rand::GetRandomFailed;
+use crate::server::ProducesTickets;
+use ring::aead;
```
Style nit: empty line between internal and external imports?

djc commented on rustls/src/crypto/ring.rs:
```diff
@@ -1,7 +1,12 @@
+use std::fmt;
```
Nit: local style is to put `std` imports below crate and external import blocks.

djc commented on rustls/src/crypto.rs:
```diff
@@ -15,3 +15,8 @@ pub trait CryptoProvider: Send + Sync + 'static {
     fn verify_equal_ct(a: &[u8], b: &[u8]) -> bool;
 }
+
+pub(crate) enum KeyExchangeError {
```
This seems arguably the wrong order to do things in? IMO this should move from `kx` to `crypto` before the contents of `kx` are moved into `crypto::ring`.

djc commented on rustls/src/lib.rs:
```diff
@@ -496,9 +495,9 @@ pub mod version {
-    pub use crate::kx::SECP256R1;
-    pub use crate::kx::SECP384R1;
-    pub use crate::kx::X25519;
+    pub use crate::crypto::ring::SECP256R1;
```
Arguably this `kx_group` module should itself move into `crypto::ring`? (Probably as a separate commit though.)

cpu replied:
I moved this in 495469b4c640e781f433441ba04d65bd0680b4a3 but left a `pub use` in the root `lib.rs`. If you were thinking we should remove all traces of this from the top-level API I can touch that up.

djc commented on rustls/src/crypto.rs:
```diff
@@ -1,5 +1,7 @@
 use crate::rand::GetRandomFailed;
 use crate::server::ProducesTickets;
+use crate::NamedGroup;
+use std::fmt::Debug;
```
Style nit: empty line between `crate` and `std` imports, please.

djc commented on rustls/src/builder.rs:
```diff
@@ -104,16 +104,21 @@ pub struct ConfigBuilder<Side: ConfigSide, State> {
     pub(crate) side: PhantomData<Side>,
 }
 
+fn friendly_names(raw_type: &str) -> (&str, &str) {
```
Style nit: should go below the `Debug` impl, instead of above it. Maybe even near the bottom of the module, if there will soon be other `Debug` impls in this module that need it?

cpu replied:
:crossed_fingers: I'm hoping I can use your idea of adding an associated type to `ConfigSide` to remove the need for this change but in the meantime I've moved it to the bottom of the module.

djc commented on rustls/src/crypto.rs:
```diff
@@ -18,8 +21,58 @@ pub trait CryptoProvider: Send + Sync + 'static {
+        name: NamedGroup,
+        supported: &[&'static Self::SupportedGroup],
+    ) -> Result<Self, KeyExchangeError>
+    where
+        Self: Sized;
```
No longer need this bound, right?

djc commented on rustls/src/crypto.rs:
```diff
@@ -18,8 +21,58 @@ pub trait CryptoProvider: Send + Sync + 'static {
+    ///
+    /// Returns an error if key generation fails.
+    fn start(skxg: &'static Self::SupportedGroup) -> Result<Self, GetRandomFailed>
+    where
+        Self: Sized;
```
Nor here.

djc commented on rustls/src/crypto.rs:
```diff
@@ -18,8 +21,58 @@ pub trait CryptoProvider: Send + Sync + 'static {
+    /// `f` is returned to the caller.
+    fn complete<T>(self, peer: &[u8], f: impl FnOnce(&[u8]) -> Result<T, ()>) -> Result<T, Error>;
+
+    /// Return all supported key exchange groups.
+    fn all_kx_groups() -> &'static [&'static Self::SupportedGroup];
```
In this case I'd move `all_kx_groups()` below `pub_key()`, since it doesn't take `&self` (sort of like a constant).

djc reviewed (commented):
Looking good! Here's a bunch of small stylistic suggestions.

djc commented:
> I did my best to find a solution that passes testing, but _there must be a better way_. Maybe there's a way to get the `CryptoProvider` parameter from the `Side` in the deeper states that I need to refer to it from?

I think we could add a `CryptoProvider` associated type to the `ConfigSide` trait?

```rust
pub trait ConfigSide: sealed::Sealed {
    type CryptoProvider: CryptoProvider;
}

impl<C: CryptoProvider> ConfigSide for crate::ClientConfig<C> {
    type CryptoProvider = C;
}
```

cpu committed:
kx: rename fields for readability/convention.

Following up on the previous commit, this commit updates the
`KeyExchange` struct's private `skxg`, `pubkey` and `privkey` fields to
be named `group`, `pub_key` and `priv_key`. This better matches the Rust
naming convention for struct members and makes for easier to understand
code.

cpu committed:
kx: move Ring key exchange impl. to crypto::ring.

This commit moves the existing Ring-based key exchange mechanisms from
`rustls/src/kx.rs` to `rustls/src/crypto/ring.rs` in anticipation of
adapting the codebase to a more general keyex trait that these types
will implement.

No changes are made to the implementation except to update import paths
to reference the new location.

cpu committed:
lib: move `kx_group` module to crypto/ring.

cpu committed:
kx: lift `KeyExchangeError` into `crypto`

The `KeyExchangeError` type is generic enough to live in the `crypto`
module. This will allow it to be shared with non-ring implementations in
the future.

cpu committed:
ring: match KeyExchange impl block to trait.

The `KeyExchange` trait's methods were ordered constructors -> complex
functions -> less complex functions. The original *ring* specific
`KeyExchange` didn't match this ordering. This commit synchronizes the
two.

cpu commented:
I believe I've updated this branch with all of your feedback. Thank you!

> I think we could add a CryptoProvider associated type to the ConfigSide trait?

Ooo, good idea. I'll give that a try shortly. My recent push hasn't included this idea yet.

cpu commented:
>> I think we could add a CryptoProvider associated type to the ConfigSide trait?
>
> Ooo, good idea. I'll give that a try shortly. 

That worked out nicely, thanks for the tip :+1: 

I dropped the commit that pulled out the Debug helper, and the commit that made some states generic w/ phantom data. It's much tidier now.  

The change to add the associated type to `Side` is integrated into the commit that introduces the KeyExchange trait throughout since that work now relies on having access to the provider.

djc commented:
(Would be nice to fix the documentation errors so CI passes on this.)

cpu commented:
> (Would be nice to fix the documentation errors so CI passes on this.)

Done!

cpu commented:
Just a note for anyone following along here: I had planned to return to this work after the 0.21.0 rustls release tag, but the priorities from my funders have changed and are now focused on client certificate improvements. I hope to return to this work after we've made progress on that front.

cpu commented:
This work was picked up by ctz in https://github.com/rustls/rustls/pull/1319
